packages/xstate-wallet/src/workflows/challenge-channel.ts

```diff
@@ -0,0 +1,62 @@
+import {Machine, StateNodeConfig, spawn, assign} from 'xstate';
+import {map, filter} from 'rxjs/operators';
+
+import {Store} from '../store';
+import {ChannelChainInfo} from '../chain';
+import {exists} from '../utils';
+
+const WORKFLOW = 'challenge-channel';
+
+export type Init = {channelId: string};
+
+export const config: StateNodeConfig<Init, any, any> = {
+  key: WORKFLOW,
+  initial: 'idle',
+  entry: 'assignChainWatcher',
+  on: {CHALLENGE_ONCHAIN_SUCCESSFULLY: 'done'},
+  states: {
+    idle: {on: {SAFE_TO_CHALLENGE: 'submit'}},
+    submit: {invoke: {src: 'submitChallengeTransaction', onDone: 'idle', onError: 'failure'}},
+    done: {type: 'final'},
+    failure: {entry: assign<any>({error: () => 'Challenge failed'})}
+  }
+};
+
+type SafeToChallenge = {type: 'SAFE_TO_CHALLENGE'};
+
+export const machine = (store: Store) => {
+  const subscribeChallengeRegisteredEvent = (ctx: Init) => {
+    return store.chain.chainUpdatedFeed(ctx.channelId).pipe(
+      map((chainInfo: ChannelChainInfo):
+        | 'CHALLENGE_ONCHAIN_SUCCESSFULLY'
+        | 'UNEXPECTED_CHALLENGE_ALREADY_EXISTS'
+        | SafeToChallenge => {
+        if (chainInfo.finalized) return 'CHALLENGE_ONCHAIN_SUCCESSFULLY';
```

Comment:
I don't follow this logic. If the channel is finalized, is there anything else to do? I think the channel should just unwind once the channel is finalized.

Reply:
Yeah this is just a placeholder right now because `FakeChain` has no notion of challenges -- it just knows `finalized` or not, and my test checks that.